gh-131146: Fix month names in a genitive case in calendar module

plashchynski opened the pull request:
The calendar module displays month names in some locales using the genitive case. This is grammatically incorrect, as the nominative case should be used when the month is named by itself. To address this issue, this change introduces new lists `alt_month_name` and `alt_month_abbr` that contain month names in the nominative case. The module now uses `%OB` format specifier to get month names in the nominative case where available.

<!-- gh-issue-number: gh-131146 -->
* Issue: gh-131146
<!-- /gh-issue-number -->

plashchynski committed:
gh-131146: Fix month names in a genitive case in calendar module.

The calendar module displays month names in some locales using the genitive case.
This is grammatically incorrect, as the nominative case should be used when the month
is named by itself. To address this issue, this change introduces new lists
`alt_month_name` and `alt_month_abbr` that contain month names in the nominative case.
The module now uses `%OB` format specifier to get month names in the nominative case
where available.

tomasr8 commented on Lib/calendar.py:
```diff
@@ -139,6 +139,24 @@ def __len__(self):
+# the month in nominative case.
+if _is_alt_mon_available():
+    alt_month_name = _localized_month('%OB')
+    alt_month_abbr = _localized_month('%Ob')
```
These will probably need a documentation entry in `calendar.rst` as well (and maybe update `month_name`/`month_abbr` to make it clear that they return the genitive case?)

tomasr8 reviewed (commented):
Thanks for the PR! This will also need a NEWS entry and tests (check `Lib/test/test_calendar.py`)

plashchynski committed:
Make `alt_month_name` and `alt_month_abbr` public in the `calendar` module.

Expose alternative month names/abbreviations so users can benefit
from alternative forms when they are available. Update documentation and
tests accordingly.

plashchynski committed:
Add NEWS entries.

picnixz commented on Doc/library/calendar.rst:
```diff
@@ -505,13 +506,39 @@ The :mod:`calendar` module exports the following data attributes:
+       >>> list(calendar.alt_month_abbr)
+       ['', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
+
+
```
```suggestion

.. data:: alt_month_name

   A sequence that represents the months of the year in the current locale
   in the grammatical form used when the month is named by itself if the locale
   provides one. If the locale does not supply an alternative form, it falls back
   to the behavior of :data:`month_name`.

       >>> import calendar
       >>> list(calendar.alt_month_name)
       ['', 'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']

   .. versionadded:: next


.. data:: alt_month_abbr

   A sequence that represents the abbreviated months of the year in the current
   locale in the grammatical form used when the month is named by itself if the
   locale provides one. If the locale does not supply an alternative form, it falls
   back to the behavior of :data:`month_abbr`.

       >>> import calendar
       >>> list(calendar.alt_month_abbr)
       ['', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']

   .. versionadded:: next


```

plashchynski replied:
Thanks for pointing out! I've added these tags in the recent commit.

picnixz commented on Lib/calendar.py:
```diff
@@ -139,6 +140,25 @@ def __len__(self):
+# On platforms that support the %OB and %Ob specifiers, it is used to get the
+# standalone form of the month name. This is required for some languages
+# such as Greek, Slavic, and Baltic languages.
+if _is_alt_mon_available():
```
```suggestion
# Check if the platform supports %OB and %Ob specifiers
def _is_alternative_month_available():
    try:
        datetime.date(2001, 1, 1).strftime('%OB')
        datetime.date(2001, 1, 1).strftime('%Ob')
    except ValueError:
        return False
    return True

# On platforms that support the %OB and %Ob specifiers, it is used to get the
# standalone form of the month name. This is required for some languages
# such as Greek, Slavic, and Baltic languages.
if _is_alternative_month_available():
```

plashchynski replied:
@picnixz will `_supports_alternative_month_names` work?

plashchynski replied:
changed to `_supports_alternative_month_names`, let me know if it won't work

picnixz commented on Lib/calendar.py:
```diff
@@ -510,9 +530,9 @@ def formatmonthname(self, theyear, themonth, withyear=True):
+            s = '%s %s' % (alt_month_name[themonth], theyear)
         else:
-            s = '%s' % month_name[themonth]
+            s = '%s' % alt_month_name[themonth]
```
What's the difference between this and `str(alt_month_name[themonth])`. And also, is it possible to have non-strings in `alt_month_name`?

plashchynski replied:
you're right, that's the same as simply `s = month_name[themonth]`

picnixz commented on Lib/test/test_calendar.py:
```diff
@@ -556,6 +557,13 @@ def test_months(self):
+        # Ensure that the alternate month names and abbreviations are equal
+        # to the regular month names and abbreviations for the "C" locale.
+        with calendar.different_locale("C"):
+            self.assertEqual(list(calendar.month_name), list(calendar.alt_month_name))
```
You can use assertListEqual as well.

plashchynski committed:
Fix py:func references in NEWS.d

plashchynski committed:
Refinement changes:

* Use assertListEqual to ensure alternate month names the same as regular
for the "C" locale.
* Use `_supports_alternative_month_names` as more descriptive name.
* Add `versionadded:: next` tags to the new APIs.

AA-Turner commented on Doc/library/calendar.rst:
```diff
@@ -493,9 +493,10 @@ The :mod:`calendar` module exports the following data attributes:
-   follows normal convention of January being month number 1, so it has a length of
-   13 and ``month_name[0]`` is the empty string.
+   A sequence that represents the months of the year in the current locale
+   in the grammatical form used when the month is part of a complete date.
```
It might be better to move these notes about grammar to a new paragraph or admonition. I fear that they are more confusing than helpful to a reader of the English docs using Python in English, as the language has no differentiation. In the same vein, the note should make it clear that this is for localised use of the module.

plashchynski replied:
@AA-Turner Will a caution like this work?

    In locales with alternative forms of month names, the :data:`month_name` sequence
    may not be suitable when a month is named by itself and not as part of a date. For instance,
    in Greek and in many Slavic and Baltic languages, :data:`month_name` will produce
    the month in genitive case. Use :data:`alt_month_name` for a form suitable for standalone use.

tomasr8 replied:
Maybe `when a month name stands by itself` instead of `when a month is named by itself`? Otherwise it sounds good.

AA-Turner commented on Doc/library/calendar.rst:
```diff
@@ -505,13 +506,43 @@ The :mod:`calendar` module exports the following data attributes:
        ['', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
 
+
+.. data:: alt_month_name
```
`alt` is not very descriptive (and could be confused for the abbreviated form). Perhaps `nominative_month_name`?

picnixz replied:
I thougzht so as well but GNU specifically uses "alternative" as terminology:

>  The %OB and %Ob formats are a C2X feature, specified in C2X to use the locale’s ‘alternative’ month name; the GNU C Library extends this specification to say that the form used in a complete date is the default and the form naming the month by itself is the alternative. 

See https://www.gnu.org/software/libc/manual/html_node/Formatting-Calendar-Time.html.

JelleZijlstra replied:
Not too familiar with all the possibilities here but not all languages where this matters might have a nominative case; e.g. some languages use an ergative-absolutive system instead.

plashchynski replied:
In the `strftime` context, it is usually referred to as an 'alternative form.' It is not always nominative, as many languages do not have a nominative form. The idea is that it a grammatical form used when the month is named by itself. In some languages it could be the nominative form or whatever else.

plashchynski replied:
May be `standalone_month_name` ?

tomasr8 replied:
I like `standalone_month_name`, it's clearer

AA-Turner commented on Lib/calendar.py:
```diff
@@ -139,6 +140,25 @@ def __len__(self):
+# such as Greek, Slavic, and Baltic languages.
+if _supports_alternative_month_names():
+    alt_month_name = _localized_month('%OB')
+    alt_month_abbr = _localized_month('%Ob')
```
Why not just use `try`/`except` here when defining the constants? I don't think the helper function is needed.

plashchynski replied:
@AA-Turner Thanks for pointing that out. I agree—the helper is redundant here, so I've removed it in the latest commit.

AA-Turner commented on Lib/test/test_calendar.py:
```diff
@@ -546,7 +546,8 @@ def test_days(self):
     def test_months(self):
-        for attr in "month_name", "month_abbr":
+        for attr in ["month_name", "month_abbr", "alt_month_name",
+                     "alt_month_abbr"]:
```
Style:
```suggestion
        for attr in ("month_name", "month_abbr", 
                     "alt_month_name", "alt_month_abbr"):
```

plashchynski committed:
Apply suggested style changes

tomasr8 commented on Doc/library/calendar.rst:
```diff
@@ -505,13 +506,43 @@ The :mod:`calendar` module exports the following data attributes:
+   A sequence that represents the months of the year in the current locale
+   in the grammatical form used when the month is named by itself if the locale
+   provides one. If the locale does not supply an alternative form, it falls back
+   to the behavior of :data:`month_name`.
```
```suggestion
   provides one. If the locale does not supply an alternative form, it is equal to :data:`month_name`.
```

tomasr8 commented on Doc/library/calendar.rst:
```diff
@@ -505,13 +506,43 @@ The :mod:`calendar` module exports the following data attributes:
+   A sequence that represents the abbreviated months of the year in the current
+   locale in the grammatical form used when the month is named by itself if the
+   locale provides one. If the locale does not supply an alternative form, it falls
+   back to the behavior of :data:`month_abbr`.
```
```suggestion
   locale provides one. If the locale does not supply an alternative form, it is equal to :data:`month_abbr`.
```

plashchynski committed:
Rename `calendar.alt_month_name` to `calendar.standalone_month_name`, and `calendar.alt_month_abbr` to `calendar.standalone_month_abbr`.

plashchynski committed:
Refine news message wording for better clarity

plashchynski committed:
Style guide compliance:

* Limit line length to 79 characters according to PEP8
* Suppress current unit links according to the Documentation Style Guide

picnixz commented on Misc/NEWS.d/next/Library/2025-03-17-21-15-04.gh-issue-131146.P9Cdx0.rst:
```diff
@@ -0,0 +1,3 @@
+Add :func:`calendar.standalone_month_name` and :func:`calendar.standalone_month_abbr` to
```
```suggestion
Add :data:`calendar.standalone_month_name` and :data:`calendar.standalone_month_abbr` to
```

plashchynski committed:
Fix roles of  and  in the NEWS.d entry.

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
-   13 and ``month_name[0]`` is the empty string.
+   A sequence that represents the months of the year in the current locale.
+   This follows normal convention of January being month number 1, so it has
+   a length of 13 and ``month_name[0]`` is the empty string.
```
This is just a formatting change and should not be included in this PR.

plashchynski replied:
It's to limit the line length to 80 characters in accordance with the [reStructuredText markup guideline](https://devguide.python.org/documentation/markup/)

StanFromIreland replied:
Again, you did not modify this section, please revert it. Existing violations do not have to be fixed, I assure you there are hundreds in the file.

plashchynski replied:
Reverted

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
-   locale.  This follows normal convention of January being month number 1, so it
-   has a length of 13 and  ``month_abbr[0]`` is the empty string.
+   locale. This follows normal convention of January being month number 1, so
+   it has a length of 13 and  ``month_abbr[0]`` is the empty string.
```
Again unrelated formatting changes

plashchynski replied:
It's to limit the line length to 80 characters in accordance with the [reStructuredText markup guideline](https://devguide.python.org/documentation/markup/).

StanFromIreland replied:
You did not modify this section, it is unrelated to this pr, please revert it.

plashchynski replied:
Reverted

StanFromIreland commented on Lib/calendar.py:
```diff
@@ -139,6 +140,17 @@ def __len__(self):
 month_abbr = _localized_month('%b')
 
+# On platforms that support the %OB and %Ob specifiers, it is used
+# to get the standalone form of the month name. This is required for
+# some languages such as Greek, Slavic, and Baltic languages.
```
```suggestion
# to get the standalone form of the month name.
```
I don't think examples are necessary here.

AA-Turner replied:
I found them useful when reviewing the PR, I don't mind them staying.

StanFromIreland commented on Lib/calendar.py:
```diff
@@ -139,6 +140,17 @@ def __len__(self):
+except ValueError:
+    # The platform does not support the %OB and %Ob specifiers.
+    standalone_month_name = month_name
+    standalone_month_abbr = month_abbr
```
```suggestion
try:
    standalone_month_name = _localized_month('%OB')
    standalone_month_abbr = _localized_month('%Ob')
except ValueError:
    standalone_month_name = month_name
    standalone_month_abbr = month_abbr
```
The previous comment covers this IMO

plashchynski replied:
I've removed the comment

StanFromIreland commented on Misc/NEWS.d/next/Library/2025-03-17-21-15-04.gh-issue-131146.P9Cdx0.rst:
```diff
@@ -0,0 +1,3 @@
+Add :data:`calendar.standalone_month_name` and :data:`calendar.standalone_month_abbr` to
+provide month names and abbreviations in a grammatical form used when a month name stands
+by itself if the locale provides one.
```
Both the NEWS entries are for the same issue, they could just be merged. "Fix ... by adding :data:`calendar.standalone_month_name` and :data:`calendar.standalone_month_abbr`. These provide ... "

plashchynski replied:
Merged

StanFromIreland commented on Lib/test/test_calendar.py:
```diff
@@ -556,6 +557,16 @@ def test_months(self):
+        # Ensure that the standalone month names and abbreviations are
+        # equal to the regular month names and abbreviations for
+        # the "C" locale.
+        with calendar.different_locale("C"):
```
Why not test a different locale like 'pl_PL' which is actually affected?

plashchynski replied:
@StanFromIreland they could be unavailable on the platform where tests run. Only "C" locale is guaranteed to be available on POSIX. I didn't find other tests where locale switching would be used.

plashchynski replied:
I've added a new test `test_standalone_month_name_and_abbr_pl_locale` to verify the names and abbrs specifically in `pl_PL`.

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
 
+   .. caution::
+
+      In locales with alternative forms of month names, the :data:`!month_name` sequence
```
```suggestion
      In locales with alternative month names forms, the :data:`!month_name` sequence
```
Simpler rewording

plashchynski replied:
Fixed, thanks.

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
+       >>> import calendar
+       >>> list(calendar.standalone_month_name)
+       ['', 'January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
+
```
This example is pointless in the English locale, and it links to `month_name` with the exact same example.

plashchynski replied:
I agree, it's not relevant for English. But which locale should I use here?

StanFromIreland replied:
We do not have to provide an example.

plashchynski replied:
Removed

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
+       >>> import calendar
+       >>> list(calendar.standalone_month_abbr)
+       ['', 'Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
+
```
Ditto

plashchynski replied:
Removed

StanFromIreland commented on Lib/calendar.py:
```diff
@@ -139,6 +140,17 @@ def __len__(self):
 month_name = _localized_month('%B')
 month_abbr = _localized_month('%b')
 
+# On platforms that support the %OB and %Ob specifiers, it is used
```
```suggestion
# On platforms that support the %OB and %Ob specifiers, they are used
```
grammar

plashchynski replied:
Thanks, fixed

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
+
+   A sequence that represents the months of the year in the current locale
+   in the grammatical form used when a month name stands by itself if the locale
+   provides one. If the locale does not supply an alternative form, it is equal to
```
```suggestion
   provides one. If the locale does not supply a standalone form, it is equal to
```
Maybe this would be clearer?

plashchynski replied:
Thanks, fixed

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
+
+   A sequence that represents the months of the year in the current locale
+   in the grammatical form used when a month name stands by itself if the locale
+   provides one. If the locale does not supply an alternative form, it is equal to
```
```suggestion
   A sequence that represents the months of the year in the current locale
   in the standalone form if the locale provides one. Else it is equivalent to
```
Much simpler format, as docs should be [e.g.](https://developer.apple.com/documentation/foundation/date/formatstyle/symbol/standalonemonth)

plashchynski replied:
Updated

StanFromIreland commented on Doc/library/calendar.rst:
```diff
@@ -493,25 +493,68 @@ The :mod:`calendar` module exports the following data attributes:
+
+   A sequence that represents the months of the year in the current locale
+   in the grammatical form used when a month name stands by itself if the locale
+   provides one. If the locale does not supply an alternative form, it is equal to
```
```suggestion
   A sequence that represents the months of the year in the current locale
   in the standalone form if the locale provides one. Else it is equivalent to
```
Much simpler format, as docs should be [e.g.](https://developer.apple.com/documentation/foundation/date/formatstyle/symbol/standalonemonth)

plashchynski replied:
Updated

plashchynski committed:
Fix grammar and wording

plashchynski committed:
Rewording

plashchynski committed:
Revert unrelated formatting changes

plashchynski committed:
Remove the examples as they are pointless in the English locale

plashchynski committed:
Remove the comment, as the previous comment covers it

plashchynski committed:
Merge NEWS entries as they both are for the same issue

plashchynski committed:
Simpler wording

plashchynski committed:
Merge branch 'main' into fix_calendar_genitive_case

plashchynski committed:
Add a test for standalone month names and abbrs in the pl_PL locale

serhiy-storchaka reviewed (approved):
LGTM. :+1:
